src/lib/utils/meta_tables/meta_system_utilization_table.cpp

```diff
@@ -1,268 +1,320 @@
+
+  return memory_usage;
+#endif
+
+  Fail("Method not implemented for this platform");
+}
+
+/**
+ * Returns a struct that contains the virtual and physical memory used by this process in bytes.
+ * - Virtual Memory is the total memory usage of the process
+ * - Physical Memory is the resident set size (RSS), the portion of memory that is held in RAM
+ */
+MetaSystemUtilizationTable::ProcessMemoryUsage MetaSystemUtilizationTable::_get_process_memory_usage() {
+#ifdef __linux__
+  std::ifstream self_status_file;
+  MetaSystemUtilizationTable::ProcessMemoryUsage memory_usage{};
+  try {
+    self_status_file.open("/proc/self/status", std::ifstream::in);
+
+    std::string self_status_line;
+    while (std::getline(self_status_file, self_status_line)) {
+      if (self_status_line.starts_with("VmSize")) {
+        memory_usage.virtual_memory = _parse_value_string(self_status_line)[0] * 1024;
+      } else if (self_status_line.starts_with("VmRSS")) {
+        memory_usage.physical_memory = _parse_value_string(self_status_line)[0] * 1024;
+      }
+    }
+
+    self_status_file.close();
+  } catch (std::ios_base::failure& fail) {
+    Fail("Failed to read /proc/self/status (" + fail.what() + ")");
+  }
+
+  return memory_usage;
+#endif
+
+#ifdef __APPLE__
+  struct task_basic_info info {};
+  mach_msg_type_number_t count = TASK_BASIC_INFO_COUNT;
+  const auto ret = task_info(mach_task_self(), TASK_BASIC_INFO, reinterpret_cast<task_info_t>(&info), &count);
+  Assert(ret == KERN_SUCCESS, "Failed to get task_info");
+
+  return {info.virtual_size, info.resident_size};
+#endif
+
+  Fail("Method not implemented for this platform");
+}
+
+/**
+ * This returns the actually allocated memory. It differs from _get_process_memory_usage in that it only returns
+ * memory that has been allocated. The reported virtual memory consumption is usually higher than the amount of
+ * allocated memory as free'd memory is not immediately returned to the system (either due to internal page
+ * fragmentation or because jemalloc keeps empty pages for future use). In rare cases, it can also be higher than
+ * the amount of virtual memory if memory has been allocated but not committed yet.
+ */
+std::optional<size_t> MetaSystemUtilizationTable::_get_allocated_memory() {
+#ifdef HYRISE_WITH_JEMALLOC
+  if (HYRISE_DEBUG) {
+    // Check that jemalloc was built with statistics support
+
+    bool stats_enabled;
+    size_t stats_enabled_size{sizeof(stats_enabled)};
+
+    auto error_code = mallctl("config.stats", &stats_enabled, &stats_enabled_size, nullptr, 0);
+    Assert(!error_code, "Cannot check if jemalloc was built with --stats_enabled");
+    Assert(stats_enabled, "Hyrise's jemalloc was not build with --stats_enabled");
+  }
+
+  // Before retrieving the statistics, we need to update jemalloc's epoch to get current values. See the mallctl
+  // documentation for details.
+  {
+    static uint64_t epoch = 1;
+    epoch++;
```

Comment:
To me, the documentation sounds like it is incrementing the epoch itself: 

> "epoch" (uint64_t) rw
If a value is passed in, refresh the data from which the mallctl* functions report values, and increment the epoch.

Reply:
Mmh, what is the purpose of the value being passed in then?

Reply:
I don't know if it has to be passed in or should be `nullptr` in that case.

Reply:
Also, I have never seen these functions before and was only taking a short look at the documentation.

Reply:
i think `epoch++` isn't needed. 
see how jemalloc uses it for it's own stats printing:
[https://github.com/jemalloc/jemalloc/blob/ea6b3e973b477b8061e0076bb257dbd7f3faa756/src/stats.c#L1412](https://github.com/jemalloc/jemalloc/blob/ea6b3e973b477b8061e0076bb257dbd7f3faa756/src/stats.c#L1412)

Reply:
Thanks for looking that up!